README.md
# exotic-industries-containers


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.0.5
Date: 04.06.2023
  Features:
    - Added new 6x6 container
---------------------------------------------------------------------------------------------------
Version: 0.0.4
Date: 02.03.2023
  Bugfixes:
    - Increased health of containers
    - Swapped remaining vanilla chests for containers in recipes
---------------------------------------------------------------------------------------------------
Version: 0.0.3
Date: 01.02.2023
  Bugfixes:
    - changed usage of steel chest in recipes to 1x1 container
---------------------------------------------------------------------------------------------------
Version: 0.0.2
Date: 03.12.2022
  Features:
    - made textures more saturated

data.lua
--====================================================================================================
--PRE INIT
--====================================================================================================

-- lib and paths

require("lib/paths")

--====================================================================================================
--MAIN CONTENT CODE
--====================================================================================================

-- add new containers

require("prototypes/containers")

-- add new subgroups
data:extend({
    {
        type = "item-subgroup",
        name = "ei_containers-1x1",
        group = "logistics",
        order = "a-a"
    },
    {
        type = "item-subgroup",
        name = "ei_containers-2x2",
        group = "logistics",
        order = "a-b"
    },
    {
        type = "item-subgroup",
        name = "ei_containers-6x6",
        group = "logistics",
        order = "a-c"
    }
})

data-updates.lua
ei_containers_lib = require("lib/lib")

--====================================================================================================
--CONTENT UPDATES
--====================================================================================================

-- reorganize vanilla data

ei_containers_lib.switch_recipe("steel-chest", "ei_1x1-container")
ei_containers_lib.switch_recipe("logistic-chest-active-provider", "ei_1x1-container_pink")
ei_containers_lib.switch_recipe("logistic-chest-passive-provider", "ei_1x1-container_red")
ei_containers_lib.switch_recipe("logistic-chest-requester", "ei_1x1-container_blue")
ei_containers_lib.switch_recipe("logistic-chest-storage", "ei_1x1-container_yellow")
ei_containers_lib.switch_recipe("logistic-chest-buffer", "ei_1x1-container_green")

-- inserter bigger containers

ei_containers_lib.add_bigger_containers()

-- fix logistic robotics
if data.raw["technology"]["logistic-robotics"] then
    table.insert(data.raw["technology"]["logistic-robotics"].effects, {
        type = "unlock-recipe",
        recipe = "ei_2x2-container_red"
    })

    table.insert(data.raw["technology"]["logistic-robotics"].effects, {
        type = "unlock-recipe",
        recipe = "ei_2x2-container_yellow"
    })
end

-- add filter versions
table.insert(data.raw["technology"]["steel-processing"].effects, {
    type = "unlock-recipe",
    recipe = "ei_1x1-container_filter"
})

table.insert(data.raw["technology"]["steel-processing"].effects, {
    type = "unlock-recipe",
    recipe = "ei_2x2-container_filter"
})

table.insert(data.raw["technology"]["steel-processing"].effects, {
    type = "unlock-recipe",
    recipe = "ei_6x6-container_filter"
})

data-final-fixes.lua
ei_containers_lib = require("lib/lib")

--====================================================================================================
--FINAL FIXES
--====================================================================================================

-- swap steel chest with 1x1 container
ei_containers_lib.swap_item_in_recipe("steel-chest", "ei_1x1-container")
ei_containers_lib.swap_item_in_recipe("logistic-chest-passive-provider", "ei_1x1-container_red")
ei_containers_lib.swap_item_in_recipe("logistic-chest-active-provider", "ei_1x1-container_pink")
ei_containers_lib.swap_item_in_recipe("logistic-chest-storage", "ei_1x1-container_yellow")
ei_containers_lib.swap_item_in_recipe("logistic-chest-requester", "ei_1x1-container_blue")
ei_containers_lib.swap_item_in_recipe("logistic-chest-buffer", "ei_1x1-container_green")

prototypes/containers.lua
ei_containers_lib = require("lib/lib")

--====================================================================================================
--1x1 CONTAINER
--====================================================================================================

ei_containers_lib.make_all(1, nil, 48, 2, false, {
    {"steel-plate", 10}
})

ei_containers_lib.make_all(1, "blue", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 3},
    {"advanced-circuit", 1}
})

ei_containers_lib.make_all(1, "red", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 3},
    {"advanced-circuit", 1}
})

ei_containers_lib.make_all(1, "pink", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 3},
    {"advanced-circuit", 1}
})

ei_containers_lib.make_all(1, "yellow", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 3},
    {"advanced-circuit", 1}
})

ei_containers_lib.make_all(1, "green", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 3},
    {"advanced-circuit", 1}
})

ei_containers_lib.make_all(1, "filter", 48, 2, false, {
    {"ei_1x1-container", 1},
    {"electronic-circuit", 1}
})

--====================================================================================================
--2x2 CONTAINER
--====================================================================================================

ei_containers_lib.make_all(2, nil, 200, 2, false, {
    {"ei_1x1-container", 4},
    {"steel-plate", 25}
})

ei_containers_lib.make_all(2, "blue", 200, 2, true, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10},
    {"advanced-circuit", 6},
    {"steel-plate", 5}
})

ei_containers_lib.make_all(2, "red", 200, 2, true, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10},
    {"advanced-circuit", 6},
    {"steel-plate", 5}
})

ei_containers_lib.make_all(2, "pink", 200, 2, true, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10},
    {"advanced-circuit", 6},
    {"steel-plate", 5}
})

ei_containers_lib.make_all(2, "yellow", 200, 2, true, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10},
    {"advanced-circuit", 6},
    {"steel-plate", 5}
})

ei_containers_lib.make_all(2, "green", 200, 2, true, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10},
    {"advanced-circuit", 6},
    {"steel-plate", 5}
})

ei_containers_lib.make_all(2, "filter", 200, 2, false, {
    {"ei_2x2-container", 1},
    {"electronic-circuit", 10}
})

--====================================================================================================
--6x6 CONTAINER
--====================================================================================================

ei_containers_lib.make_all(6, nil, 1000, 2, false, {
    {"ei_2x2-container", 4},
    {"steel-plate", 50},
    {"stone-brick", 25}
})

ei_containers_lib.make_all(6, "blue", 1000, 2, true, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20},
    {"advanced-circuit", 10},
    {"steel-plate", 10}
})

ei_containers_lib.make_all(6, "red", 1000, 2, true, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20},
    {"advanced-circuit", 10},
    {"steel-plate", 10}
})

ei_containers_lib.make_all(6, "pink", 1000, 2, true, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20},
    {"advanced-circuit", 10},
    {"steel-plate", 10}
})

ei_containers_lib.make_all(6, "yellow", 1000, 2, true, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20},
    {"advanced-circuit", 10},
    {"steel-plate", 10}
})

ei_containers_lib.make_all(6, "green", 1000, 2, true, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20},
    {"advanced-circuit", 10},
    {"steel-plate", 10}
})

ei_containers_lib.make_all(6, "filter", 1000, 2, false, {
    {"ei_6x6-container", 1},
    {"electronic-circuit", 20}
})

lib/lib.lua
local ei_containers_lib = {}

--====================================================================================================
--BASE PROTOTYPES
--====================================================================================================

local entity_base = {
    name = "ei_1x1-container",
    type = "container",
    flags = {"placeable-neutral", "player-creation"},
    icon = ei_containers_item_path.."1x1-container.png",
    icon_size = 64,
    selection_box = {{-0.5, -0.5}, {0.5, 0.5}},
    collision_box = {{-0.35, -0.35}, {0.35, 0.35}},
    picture = {
        layers = {
            {
                filename = ei_containers_entity_path.."1x1-container.png",
                priority = "extra-high",
                width = 256*2,
                height = 256*2,
                scale = 0.13,
            },
            {
                filename = ei_containers_entity_path.."1x1-container_shadow.png",
                priority = "extra-high",
                width = 256*2,
                height = 256*2,
                scale = 0.13,
                draw_as_shadow = true,
            }
        }
    },
    minable = {
        mining_time = 1,
        result = "ei_1x1-container"
    },
    inventory_size = 48,
    circuit_connector_sprites = data.raw["container"]["steel-chest"].circuit_connector_sprites,
    circuit_wire_connection_point = data.raw["container"]["steel-chest"].circuit_wire_connection_point,
    circuit_wire_max_distance = data.raw["container"]["steel-chest"].circuit_wire_max_distance,
    -- inventory_type = "with_filters_and_bar",
    scale_info_icons = true,
    fast_replaceable_group = "container",
    max_health = 350,
}

local item_base = {
    name = "ei_1x1-container",
    type = "item",
    icon = ei_containers_item_path.."1x1-container.png",
    icon_size = 64,
    subgroup = "energy",
    order = "f[nuclear-energy]-b[basic-heat-pipe]",
    place_result = "ei_1x1-container",
    stack_size = 50
}

local recipe_base = {
    name = "ei_1x1-container",
    type = "recipe",
    category = "crafting",
    energy_required = 1,
    ingredients =
    {
        {"steel-plate", 1},
    },
    result = "ei_1x1-container",
    result_count = 1,
    enabled = false,
    always_show_made_in = true,
    main_product = "ei_1x1-container",
}

local order_dict = {
    ["none"] = "a",
    ["filter"] = "b",
    ["blue"] = "e",
    ["red"] = "c",
    ["yellow"] = "d",
    ["green"] = "f",
    ["pink"] = "g",
}


--====================================================================================================
--UTIL FUNCTIONS
--====================================================================================================

function ei_containers_lib.switch_recipe(old_recipe, new_recipes)
    -- loop over all techs and look into their effects
    -- if the effect is a recipe unlock, check if it is the old recipe
    -- if it is, replace it with the new recipe

    for _, tech in pairs(data.raw.technology) do
        if tech.effects then
            for _, effect in pairs(tech.effects) do
                if effect.type == "unlock-recipe" and effect.recipe == old_recipe then
                    effect.recipe = new_recipes
                end
            end
        end
    end
end


function ei_containers_lib.swap_item_in_recipe(item, new_item)
    -- loop over all recipes and look into their ingredients
    -- if the ingredient is the item, replace it with the new item

    for _, recipe in pairs(data.raw.recipe) do
        -- account normal and expensive mode aswell
        for _, difficulty in pairs({recipe.expensive or nil, recipe.normal or nil, recipe}) do
            if difficulty.ingredients then
                for _, ingredient in pairs(difficulty.ingredients) do
                    if ingredient[1] == item then
                        ingredient[1] = new_item
                    end
                end
            end
        end
    end

end


function ei_containers_lib.add_bigger_containers()
    -- loop over all techs and look into their effects
    -- if the effect is a recipe unlock, check if it unlocks a 1x1_container
    -- if so also add unlock for the 2x2 and 6x6 container

    local done = {}

    for i, tech in pairs(data.raw.technology) do
        if tech.effects then
            for j, effect in pairs(tech.effects) do
                if effect.type == "unlock-recipe" then
                    -- log
                    -- log("effect: "..effect.recipe)

                    start, stop = string.find(effect.recipe, "ei_1x1-")

                    if start and stop then
                        -- log 
                        log("Adding bigger container to tech: "..tech.name)

                        -- swap ei_1x1- with ei_2x2-
                        local new_recipe_2x2 =  "ei_2x2"..string.sub(effect.recipe, 7)
                        local new_recipe_6x6 =  "ei_6x6"..string.sub(effect.recipe, 7)

                        -- dont do stuff multiple times
                        if not done[new_recipe_2x2] then
                            table.insert(data.raw.technology[i].effects, {type = "unlock-recipe", recipe = new_recipe_2x2})
                            table.insert(data.raw.technology[i].effects, {type = "unlock-recipe", recipe = new_recipe_6x6})

                            done[new_recipe_2x2] = true
                        end
                    end
                end
            end
        end
    end

end


function ei_containers_lib.make_all(size, typus, slots, time, animation, ingredients)
    -- build item, recipe, entity
    ei_containers_lib.make_recipe(size, typus, ingredients, time)
    ei_containers_lib.make_item(size, typus)
    ei_containers_lib.make_container(size, slots, typus, animation)
end

function ei_containers_lib.make_recipe(size, typus, ingredients, time)
    local recipe = table.deepcopy(recipe_base)

    if typus then
        typename = "_"..typus
    else
        typename = ""
    end

    name = size.."x"..size.."-container"
    fullname = "ei_"..name..typename

    recipe.name = fullname

    recipe.ingredients = ingredients
    recipe.energy_required = time

    recipe.result = fullname
    recipe.main_product = fullname
    
    data:extend({recipe})
end


function ei_containers_lib.make_item(size, typus)
    local item = table.deepcopy(item_base)

    if typus then
        typename = "_"..typus
    else
        typename = ""
    end

    name = size.."x"..size.."-container"
    fullname = "ei_"..name..typename

    item.name = fullname
    item.place_result = fullname

    item.icon = ei_containers_item_path..name..typename..".png"

    item.subgroup = "ei_containers-"..size.."x"..size
    
    if not typus then
        typus = "none"
    end
    item.order = order_dict[typus]

    data:extend({item})
end


function ei_containers_lib.make_container(size, slots, typus, animation)
    -- size can be 1 for 1x1, 2 for 2x2, 3 for 3x3, etc.
    -- type can be blue, red, pink, filter, green, yellow

    local container = table.deepcopy(entity_base)

    if typus then
        typename = "_"..typus
    else
        typename = ""
    end
    
    -- naming
    name = size.."x"..size.."-container"
    fullname = "ei_"..name..typename

    container.name = fullname
    container.icon = ei_containers_item_path..name..typename..".png"

    image_size = 512
    adjust = 1

    if size > 2 then
        image_size = 1024
        adjust = 0.5
    end

    -- size
    container.selection_box = {{-size/2, -size/2}, {size/2, size/2}}
    container.collision_box = {{-size/2+0.15, -size/2+0.15}, {size/2-0.15, size/2-0.15}}

    -- picture
    container.picture.layers[1].filename = ei_containers_entity_path..name..typename..".png"
    container.picture.layers[1].width = image_size
    container.picture.layers[1].height = image_size

    container.picture.layers[2].filename = ei_containers_entity_path..name.."_shadow.png"
    container.picture.layers[2].width = image_size
    container.picture.layers[2].height = image_size

    container.picture.layers[1].scale = 0.13*size*adjust
    container.picture.layers[2].scale = 0.13*size*adjust

    -- inventory
    container.minable.result = fullname
    container.inventory_size = slots

    -- animation
    if animation then
        container.animation = {
            layers = {
                {
                    filename = ei_containers_entity_path..name..typename..".png",
                    priority = "extra-high",
                    width = image_size,
                    height = image_size,
                    scale = 0.13*size*adjust,
                    frame_count = 1,
                    line_length = 1,
                    animation_speed = 1,
                    repeat_count = 5,
                },
                {
                    filename = ei_containers_entity_path..name.."_beam.png",
                    priority = "extra-high",
                    width = image_size,
                    height = image_size,
                    scale = 0.13*size*adjust,
                    frame_count = 5,
                    line_length = 5,
                    animation_speed = 1,
                    run_mode = "backward",
                },
                {
                    filename = ei_containers_entity_path..name.."_shadow.png",
                    priority = "extra-high",
                    width = image_size,
                    height = image_size,
                    scale = 0.13*size*adjust,
                    draw_as_shadow = true,
                    frame_count = 1,
                    line_length = 1,
                    animation_speed = 1,
                    repeat_count = 5,
                }
            }
        }
    end

    -- for different types of containers
    if typus then
        if typus == "filter" then
            container.inventory_type = "with_filters_and_bar"
            -- also only 18 slots
            container.inventory_size = 18
        end
        -- TODO add animations

        -- different logistic versions
        if typus == "blue" then
            container.type = "logistic-container"
            container.logistic_mode = "requester"
            container.max_logistic_slots = 1000
        end

        if typus == "yellow" then
            container.type = "logistic-container"
            container.logistic_mode = "storage"
            container.max_logistic_slots = 1
        end

        if typus == "red" then
            container.type = "logistic-container"
            container.logistic_mode = "passive-provider"
            container.max_logistic_slots = 0
        end

        if typus == "green" then
            container.type = "logistic-container"
            container.logistic_mode = "buffer"
            container.max_logistic_slots = 1000
        end

        if typus == "pink" then
            container.type = "logistic-container"
            container.logistic_mode = "active-provider"
            container.max_logistic_slots = 0
        end
    end

    data:extend({container})
end



return ei_containers_lib


